addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/http/ComparableResponse.java

```diff
@@ -0,0 +1,527 @@
+        if (response1.headers.isEmpty() && response2.headers.isEmpty()) {
+            return 1; // Both empty, equal
+        } else if (response1.headers.isEmpty() || response2.headers.isEmpty()) {
+            return 0; // One empty, very different
+        }
+        float equalHeadersContent = 0;
+        int r1HeaderCount = response1.headers.keySet().size();
+        int r2HeaderCount = response2.headers.keySet().size();
+        float numberHeaders = r1HeaderCount > r2HeaderCount ? r1HeaderCount : r2HeaderCount;
+        Map<String, String> larger = response1.headers;
+        Map<String, String> smaller = response2.headers;
+        if (r2HeaderCount > r1HeaderCount) {
+            larger = response2.headers;
+            smaller = response1.headers;
+        }
+
+        for (String key : larger.keySet()) {
+            if (smaller.containsKey(key)
+                    && (smaller.get(key).equals(larger.get(key))
+                            || DYNAMIC_HEADERS.contains(key))) {
+                equalHeadersContent += 1;
+            }
+        }
+        // Similar to (equalHeaders / numberHeaders * equalHeadersContent / equalHeaders)* 0.25 +
+        // 0.75;
+        return equalHeadersContent / numberHeaders;
+    }
+
+    /**
+     * Return a heuristic evaluation between Body structure of the 2 responses. 0 means very
+     * different and 1 very similar.
+     *
+     * @param response1 The first response to compare
+     * @param response2 The second response to compare
+     */
+    public static float bodyTreesStructureHeuristic(
+            ComparableResponse response1, ComparableResponse response2) {
+        String r1ContentType = response1.getHeaders().get(HttpHeader.CONTENT_TYPE);
+        String r2ContentType = response2.getHeaders().get(HttpHeader.CONTENT_TYPE);
+
+        if (r1ContentType != null && r2ContentType != null) {
+            // We just want the type
+            // TODO should I consider different if they have different parameters?
+            // as "text/html; charset=utf-8" and "text/html
+            r1ContentType = r1ContentType.toLowerCase().split(";")[0];
+            r2ContentType = r2ContentType.toLowerCase().split(";")[0];
+
+            if (!r1ContentType.equals(r2ContentType)) {
+                return 0;
+            } else if (r1ContentType.contains("text/html") || r1ContentType.contains("json")) {
+                HashMap<String, Integer> response1paths =
+                        (HashMap<String, Integer>) response1.getAllPaths();
+                HashMap<String, Integer> response2paths =
+                        (HashMap<String, Integer>) response2.getAllPaths();
+
+                if (response1.numPaths == 0 || response2.numPaths == 0) {
+                    if (response1.numPaths != 0 || response2.numPaths != 0) {
+                        return 0;
+                    } else {
+                        return 1;
+                    }
+                }
```

Comment:
No, it can return 1 if they are both zero - the first condition passes but the second one doesnt.
But its not easy to read - maybe not nesting the ifs would help?
```
if (response1.numPaths == 0 && response2.numPaths == 0) {
  return 1;
} else if (response1.numPaths == 0 || response2.numPaths == 0) {
  return 0;
}
```
Thats always assuming that is what the code is intended to do ;)

Reply:
Yup that's much easier.

Reply:
```java
if (response1.numPaths == 0 || response2.numPaths == 0) {
      if (response1.numPaths == response2.numPaths) {
          return 1;
      } else {
          return 0;
      }
  }
```
Maybe this is more clean

Reply:
Thank you both, I need to read through the rest of the method, but one of those definitely helps.